feature/4493 Improve Evernote Document Loader

MikeMcGarry opened the pull request:
# Improve Evernote Document Loader

When exporting from Evernote you may export more than one note. Currently the Evernote loader concatenates the content of all notes in the export into a single document and only attaches the name of the export file as metadata on the document.

This change ensures that each note is loaded as an independent document and all available metadata on the note e.g. author, title, created, updated are added as metadata on each document. 

It also uses an existing optional dependency of `html2text` instead of `pypandoc` to remove the need to download the pandoc application via `download_pandoc()` to be able to use the `pypandoc` python bindings.

Fixes #4493 

## Before submitting

<!-- If you're adding a new integration, include an integration test and an example notebook showing its use! -->

## Who can review?

@eyurtsev / @dev2049

Community members can review the PR once tests pass. Tag maintainers/contributors who might be interested:

<!-- For a quicker response, figure out the right person to tag with @

        @hwchase17 - project lead

        Tracing / Callbacks
        - @agola11

        Async
        - @agola11

        DataLoaders
        - @eyurtsev

        Models
        - @hwchase17
        - @agola11

        Agents / Tools / Toolkits
        - @vowelparrot
        
        VectorStores / Retrievers / Memory
        - @dev2049
        
 -->

MikeMcGarry commented on pyproject.toml:
```diff
@@ -109,6 +109,8 @@ responses = "^0.22.0"
 pytest-asyncio = "^0.20.3"
 lark = "^1.1.5"
 pytest-mock  = "^3.10.0"
+lxml = "^4"
+html2text = "^2020.1.16"
```
Required to run the unit tests. `html2text` is currently an optional dependency for langchain and have added lxml as another optional dependency.

eyurtsev replied:
I'm still in the process of updating the testing workflow so we can handle optional dependencies, so nothing is document yet (sorry about that! :() 

Adding the dependencies here, will make the tests pass, but won't actually work when the langchain package is installed for users (since it's installed without optional dependencies).

Could you move these to the `extended_testing` group below?

```
extended_testing = ["pypdf", "pdfminer.six"]
```

MikeMcGarry replied:
Ahhh yes, no trouble at all. That sounds great, this way we don't have duplicated dependency versions in `pyproject.toml` either. I've made this change and added both `lxml` and `html2text` to `extended_testing`.

Commit:
feature/4493 Improved the Evernote loader to return a document per note instead of a single document for the entire export. Also ensured that all available metadata is added to each note

eyurtsev commented on langchain/document_loaders/evernote.py:
```diff
@@ -5,78 +5,110 @@
-        text = _parse_note_xml(self.file_path)
-        metadata = {"source": self.file_path}
-        return [Document(page_content=text, metadata=metadata)]
+        """Load documents from EverNote export file."""
```
I don't know if any users were relying on getting a single document back as a feature, if so, this would break backwards compatibility.

What do you think about adding an extra named argument in the initializer that sets the mode?

MikeMcGarry replied:
Yeah no trouble at all, the `EverNoteLoader` now takes a `load_single_document` argument in initialisation and this defaults to `True`. I think in the future we should make the default `False`, perhaps we can do this with a more significant version update.

eyurtsev commented on langchain/document_loaders/evernote.py:
```diff
@@ -5,78 +5,110 @@
+        return rsc_dict
+
+    @classmethod
+    def _parse_note(cls, note: List, prefix: Optional[str] = None) -> dict:
```
It's much more common for classmethods to return instances of the class. 

Maybe make it into a staticmethod or else detach from class as python is a land of verbs. :)

MikeMcGarry replied:
Yeah you are spot on, not sure what I was thinking, I've moved these to static methods which they should have been in the first place.

eyurtsev commented on langchain/document_loaders/evernote.py:
```diff
@@ -5,78 +5,110 @@
+            add_prefix(key): value for key, value in note_dict.items()
+        }
+
+    @classmethod
```
What do you think about using standalone functions instead? 

A developer will not have to worry about anything unexpected happening during inheritance (where identity of `cls` changes)

MikeMcGarry replied:
Yes, good point. These should be static methods instead of class methods. As they only exist to support the EverNoteLoader class I think they should exist inside the class rather than outside in the file. Are you happy with them as static methods?

MikeMcGarry replied:
I see your message above about using static methods instead so that's perfect. Have updated to be static methods.

eyurtsev commented on tests/unit_tests/document_loader/test_evernote_loader.py:
```diff
@@ -0,0 +1,75 @@
+        current_dir = pathlib.Path(__file__).parent
+        return os.path.join(current_dir, "sample_documents", notebook_name)
+
+    def test_evernoteloader_loadnotebook_eachnoteisindividualdocument(self) -> None:
```
@MikeMcGarry you can use the following (still undocumented) custom marker to get the tests to run. They'll need to be registered in the `extended_testing` extra

```suggestion
    @pytest.mark.requires("lxml", "html2txt")
    def test_evernoteloader_loadnotebook_eachnoteisindividualdocument(self) -> None:
```

MikeMcGarry replied:
Yeah sure thing, however when I add this decorator and call `make test` (after calling `poetry lock --no-update` and `poetry install -E extended_testing` all the tests are skipped?

MikeMcGarry replied:
Ahh never mind, I see how it works, had a type in `html2txt`. Fixed this. I see that it will skip the tests if the dependency is not installed. Nice!

Commit:
feature/4493 Added additional test coverage, fixed the management of optional dependencies and updated to use static methods instead of class methods

Commit:
feature/4493 Preserve existing behaviour and make the new behaviour optional. Update Jupyter Notebook example

MikeMcGarry commented:
Hi @eyurtsev thanks for your detailed review, your guidance on setting up the dependencies appropriately was really helpful and you are on the money about those methods which should be static methods rather than class methods. I've addressed all your feedback including preserving the existing behaviour and making the new behaviour available by passing `load_single_document=False` into the initialisation of the `EverNoteLoader`. I've also added some additional unit tests to improve our coverage.

eyurtsev commented:
Thank you! as a heads up, will commandeer tomorrow to help resolve merge conflicts around poetry.lock files, there's a bunch of PRs that are bumping into merge conflicts, so will be doing the same in all to merge the code faster.

Commit:
feature/4493 Run black across file

MikeMcGarry commented:
I've also rebased onto master.

MikeMcGarry commented on poetry.lock:
```diff
@@ -9994,18 +10087,18 @@ cffi = {version = ">=1.11", markers = "platform_python_implementation == \"PyPy\
 [extras]
-all = ["anthropic", "cohere", "openai", "nlpcloud", "huggingface_hub", "jina", "manifest-ml", "elasticsearch", "opensearch-py", "google-search-results", "faiss-cpu", "sentence-transformers", "transformers", "spacy", "nltk", "wikipedia", "beautifulsoup4", "tiktoken", "torch", "jinja2", "pinecone-client", "pinecone-text", "weaviate-client", "redis", "google-api-python-client", "wolframalpha", "qdra ...
-azure = ["azure-identity", "azure-cosmos", "openai", "azure-core"]
+all = ["O365", "aleph-alpha-client", "anthropic", "arxiv", "atlassian-python-api", "azure-cosmos", "azure-identity", "beautifulsoup4", "clickhouse-connect", "cohere", "deeplake", "docarray", "duckduckgo-search", "elasticsearch", "faiss-cpu", "google-api-python-client", "google-search-results", "gptcache", "hnswlib", "html2text", "huggingface_hub", "jina", "jinja2", "jq", "lancedb", "lark", "lxml" ...
+azure = ["azure-core", "azure-cosmos", "azure-identity", "openai"]
```
When I ran `poetry lock --no-update` it put all the back in alphabetical order.

dev2049 commented:
seems like one of the test files is 11k lines, any chance we could use a smaller one 😅

MikeMcGarry commented:
> seems like one of the test files is 11k lines, any chance we could use a smaller one 😅

dev2049 commented:
> It's a base64 encoded image. I wanted to make sure images were removed from the note and didn't end up in the page context.
> 
> Do we have some size constraints we want to consider with test data files? These aren't part of the package contents when published to PyPi right?

they're not part of the package, it's still nice to keep the repo small though. could we just use a small image (shouldn't change quality of the test right?)

Commit:
Use smaller image for media test and move to new document_loaders directory

MikeMcGarry commented:
> > It's a base64 encoded image. I wanted to make sure images were removed from the note and didn't end up in the page context.
> > Do we have some size constraints we want to consider with test data files? These aren't part of the package contents when published to PyPi right?
> 
> they're not part of the package, it's still nice to keep the repo small though. could we just use a small image (shouldn't change quality of the test right?)

Okay sure thing, I've scaled down the image it's now 18KB instead of 2.1MB.

MikeMcGarry commented:
@dev2049 / @eyurtsev I've accommodated all the feedback and rebased onto master. Are there any outstanding points we need to address?

Commit:
Remove whitespace changes

eyurtsev commented:
Code looks ready for merging

dev2049 committed:
fmt

dev2049 committed:
lint
